Docs/polytopes

tbrysiewicz opened the pull request:
More documentation for the polyhedral geometry side of Oscar.
Some moderate code changes.
Ability to save/load objects.

alexej-jordan committed:
merged with stashed changes

alexej-jordan committed:
added function for pretty printing constraints

alexej-jordan committed:
adjusted docstrings for polyhedron

alexej-jordan committed:
merged current master and resolved conflicts

alexej-jordan committed:
added and adjusted docstrings for cones, fans and polyhedra

alexej-jordan committed:
Merge branch 'master' into Polymake/docs-and-small-user-functions

micjoswig committed:
doc edits for polhedral geometry

micjoswig committed:
typo

micjoswig committed:
Merge branch 'master' into docs/Polytopes

alexej-jordan committed:
midwork commit: docstrings for linear programs and documentary page cones

alexej-jordan committed:
Merge branch 'master' into Polymake/docs-and-small-user-functions

alexej-jordan committed:
include cones page; adjusted header in docs block for polyhedra

tbrysiewicz committed:
Setup for populating cone and fan documentation

tbrysiewicz committed:
Setup for cone and fan doc

tbrysiewicz committed:
Added some doc for fans and cones

alexej-jordan committed:
Merge branch 'master' into Polymake/docs-and-small-user-functions

alexej-jordan committed:
merged branch Polymake/small-user-functions-and-docs into docs/Polytope and resolved conflicts

alexej-jordan committed:
fixed header in docs block

tbrysiewicz committed:
Added section on other constructions

tbrysiewicz committed:
sketched out linear program doc. Functions still need documentation

tbrysiewicz committed:
Restructured polyhedral geometry docs: polyhedra constructions was too long

tbrysiewicz committed:
renamed old polyhedra file to avoid editing issues

tbrysiewicz committed:
Eliminated old files and structure, can revive if necessary

benlorenz committed:
Merge remote-tracking branch 'origin/master' into docs/Polytopes

micjoswig committed:
docstring feasible_region

micjoswig committed:
capitalizing a doc chapter heading

micjoswig committed:
differences between Oscar and polymake

tbrysiewicz committed:
Added running example in LP documentation. Added LP function docs and examples

tbrysiewicz committed:
Added Cartesian products and convex hulls of pairs of polyhedra

lkastner committed:
Fix minor typo bugs in serialization

lkastner committed:
Document serialization for cones and polyhedral fans

lkastner committed:
Merge branch 'docs/Polytopes' of github.com:oscar-system/Oscar.jl into docs/Polytopes

micjoswig committed:
LP doc minor edit: sync variables in docstring and surrounding text

micjoswig committed:
DEV update

micjoswig committed:
polymake doc update

tbrysiewicz committed:
fixed LP test

lkastner committed:
Polytopes: Add doc on serializing polyhedra

lkastner committed:
Polytopes: Store convention as attachment of polymake LP for serialization

lkastner committed:
Polytopes: Add serialization for LinearProgram

lkastner committed:
Polytopes: Document serialization of LinearProgram

benlorenz commented on docs/doc.main:
```diff
@@ -19,7 +19,6 @@
 
   "Rings" => [
      "Rings/intro.md",
-     "AbstractAlgebra/ring.md",
```
Was this on purpose? (the [commit](https://github.com/oscar-system/Oscar.jl/commit/ebbe14bec5902e948af1719cbc3aa876a0a9a1ea) also contains no indication why this was removed)

tbrysiewicz replied:
I was getting an error that no such file existed so I just deleted the line with the intention of putting it back. Sorry.

benlorenz commented on docs/doc.main:
```diff
@@ -37,7 +36,6 @@
 
   "Fields" => [
      "Fields/intro.md",
-     "AbstractAlgebra/field.md",
```
Also on purpose?

tbrysiewicz replied:
I was getting an error that no such file existed so I just deleted the line with the intention of putting it back. Sorry.

benlorenz commented on docs/src/PolyhedralGeometry/pg_fans.md:
```diff
@@ -0,0 +1,43 @@
+
+## Construction
+
+To construct a polyhedral fan, you must pass the rays of each cone in the fan, along with an `IncidenceMatrix` encoding which rays are extremal with respect to which cones.
```
Redundant and non-extremal rays are also allowed in the input, the rays for each cone just need to be a generating set (maybe refer to `positive_hull` for `Cone`).

benlorenz commented on src/Polytopes/Cone/properties.jl:
```diff
@@ -136,12 +121,9 @@ codim(C::Cone) = ambient_dim(C)-dim(C)
 
-# Arguments
-- `C::Cone`: A cone.
+Determine whether `C` is pointed, i.e. whether 0 is a face of `C`.
```
```suggestion
Determine whether `C` is pointed, i.e. whether the origin is a face of `C`.
```

benlorenz commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -71,30 +112,65 @@ Base.length(iter::MaximalConeIterator) = nmaximal_cones(iter.PF)
+Return the dimension of `PF`.
+
+# Examples
+This fan in the plane contains a 2-dimensional cone and is this 2-dimensional
```
```suggestion
This fan in the plane contains a 2-dimensional cone and is thus 2-dimensional
```

benlorenz commented on src/Polytopes/Polyhedron/standard_constructions.jl:
```diff
@@ -114,16 +104,13 @@ end
 # Examples
-The positive orthant of the plane is the intersection of the two halfspaces with $x>0$ and $y>0$ respectively.
+The positive orthant of the plane is the intersection of the two halfspaces with
+$x>0$ and $y>0$ respectively.
```
```suggestion
$x≥0$ and $y≥0$ respectively.
```

benlorenz commented on src/Polytopes/PolyhedralFan/constructors.jl:
```diff
@@ -5,19 +5,61 @@
+`Rays` is given row-wise as representative vectors, with lineality generated by
+the rows of `LS`. The cones are given as an IncidenceMatrix, where the columns
+represent the rays and the rows represent the cones. There is a `1` at position
+`(i,j)` if cone `i` has ray `j` as extremal ray, otherwise there is a `0`.
```
```suggestion
`(i,j)` if cone `i` has ray `j` as a generator, otherwise there is a `0`.
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,103 @@
+
+# Constructions
+
+The standard way to define a polyhedron is by either giving a V-description or an H-description.
```
What's a H-/V-description? Is it the same as H-/V-representation? If yes, why two names? If no, what is what and how do they relate?

Also: perhaps $H$ and $V$ instead of H and V?

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,103 @@
+
+### Intersecting halfspaces: H-description
+
+Based on the definition of an $H$-representation, the constructor of `Polyhedron` can be used:
```
What's the purpose of this kind of filler text? Aren't the docstrings being inserted here plus the headline explanation enough? If not, perhaps those should be improved instead of inserting chit-chat?

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+Polyhedron(A::Union{Oscar.MatElem,AbstractMatrix}, b)
+```
+
+The following defines a polytope, as a `Polyhedron`, via a $V$-representation by calling the following function:
````
If it is about V-reps, why is it in the section on H-description?

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,103 @@
+
+### Computing convex hulls: V-description
+
+This is a standard triangle, defined via a (redundant) $V$-representation  and its unique minimal $H$-representation:
```
Is this supposed to be an example, or what? For what, `convex_hull`? Then perhaps it should be in the docstring of `convex_hull` instead?

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,103 @@
+
+
+## Named polyhedra
+Some commonly used polyhedra are able to be constructed by name. For example
```
I don't think polyhedra "are able" to construct anything. Also, this is a reference manual, so why "for example"? Don't you list all constructors we provide here? If not, why not?

Personally, I'd just drop this sentence completely (like many of the others, although I won't repeat this each time). But if you really want to keep it, how about this:
```suggestion
For some commonly used types of polyhedra, dedicated constructors are provided:
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+cross
+```
+
+```@docs
````
```suggestion
cube
cross
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+```
+
+## Operations on polyhedra
+Polyhedra can be produced through operations on other polyhedra. For example, they can be added using Minkowski addition or scaled; each of which results in a new polyhedron.
````
Again, I would just remove this...

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+intersect(::Polyhedron, ::Polyhedron)
+```
+
+```@docs
````
```suggestion
*(::Int, ::Polyhedron)
intersect(::Polyhedron, ::Polyhedron)
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+*(::Polyhedron, ::Polyhedron)
+```
+
+The convex hull of two polytopes is supported by the function `convex_hull`.
````
```suggestion
The convex hull of two polytopes can be computed via `convex_hull`.
```
... or just drop it, the docstring should speak for itself.... (If it doesn't, it should be improved accordingly)

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,103 @@
+
+## Polyhedra from other objects
+
+There is a growing list of functionality for constructing polyhedra coming from other mathematical objects. We list a handful below.
```
This would be fine for a presentation, a demo, a press release, but not a reference manual. Here, we talk about what is, not what might be (as "growing" does), and we strive to list all relevant functions we provide, not just "a handful".

So I'd remove this

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+orbit_polytope
+```
+
+```@docs
````
```suggestion
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_intro.md:
```diff
@@ -0,0 +1,29 @@
+
+# Introduction
+
+Let $\mathbb{F}$ be an ordered field; the most prominent case here is where $\mathbb{F}=\mathbb{Q}$ are the rational numbers.
```
"the most prominent case" -- wha it that supposed to mean? Prominent with whom and for what, and why is it relevant for the reader?

Perhaps you wanted something like this?
```suggestion
Let $\mathbb{F}$ be an ordered field, for example the field $\mathbb{Q}$ of rational numbers.
```

fingolfin replied:
Also, this text seems to replicate things that were started in PR #508 -- how do that PR and the one here relate to each other?

lkastner replied:
"Prominent" just means that this is the field that is used most. Functionality for other fields is limited. Most implementations and most software we interface only works with QQ.
@alexej-jordan Can you maybe comment on your PR?

alexej-jordan replied:
Within a common zoom session we merged that branch into this docs branch. I will close #508.

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_intro.md:
```diff
@@ -0,0 +1,29 @@
+A set $P \subseteq \mathbb{F}^n$ is called a (convex) polyhedron if it can be written as the intersection of finitely many closed affine halfspaces in $\mathbb{F}^n$.
+That is, there exists a matrix $A$ and a vector $b$ such that
+$$P = P(A,b) = \{ x \in \mathbb{F}^n \mid Ax \leq b\}.$$
+Writing $P$ as above is called an $H$-representation of $P$.
```
OK, so it seems my comments for PR #508 might apply to this PR here as well?
```suggestion
Writing $P$ as above is called an *$H$-representation* of $P$.
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_intro.md:
```diff
@@ -0,0 +1,29 @@
+
+When a polyhedron $P \subset \mathbb{F}^n$ is bounded, it is called a polytope and the fundamental theorem of polytopes states that it may be written as the convex hull of finitely many points.
+That is $$P = \textrm{conv}(p_1,\ldots,p_N), p_i \in \mathbb{F}^n.$$
+Writing $P$ in this way is called a $V$-representation.
```
```suggestion
Writing $P$ in this way is called a *$V$-representation*.
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,103 @@
+
+This is a standard triangle, defined via a (redundant) $V$-representation  and its unique minimal $H$-representation:
+
+```@repl oscar
````
Please consider using `jldoctest` instead of `@repl` in as many cases as possible (ideally, all).

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
```diff
@@ -0,0 +1,55 @@
+
+# `Polyhedron` and `polymake`'s `Polytope`
+
+Many polyhedral computations are done through `polymake`.
```
The user may not know what "polymake" is at this point. I'd suggest to at least insert one brief sentence explaining it, and perhaps a reference to (to be added) entry in the bibliography for polymake, and/or a link to the polymake homepage?

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
```diff
@@ -0,0 +1,55 @@
+
+Many polyhedral computations are done through `polymake`.
+This is visible in the structure `Polyhedron` via a pointer `pm_polytope` to the corresponding `polymake` object.
+This allows to apply all functionality of `polymake` by calling a suitable `Polymake.jl` function on this pointer.
```
Again, the read may not know anything about `Polymake.jl` here. So perhaps also introduce that?

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
```diff
@@ -0,0 +1,55 @@
+This is visible in the structure `Polyhedron` via a pointer `pm_polytope` to the corresponding `polymake` object.
+This allows to apply all functionality of `polymake` by calling a suitable `Polymake.jl` function on this pointer.
+
+To allow both `Oscar`'s and `polymake`'s functionality to be applicable to a polyhedron object, it can be converted back and forth:
```
I think that's a rather unfortunate way to state things, given that Oscar is built on polymake, via Polymake.jl, while this makes it sound as if they were two separate systems. 

Also, I hope these are not really doing conversions, but rather are about wrapping and unwrapping objects  Polymake.jl.

So, perhaps like this:
```suggestion
Sometimes it can be necessary to directly invoke some `polymake` functions on an Oscar `Polyhedron` object (e.g. because some functionality has not yet been made available via Oscar's interface). In that case, the following two functions allow extracting the underlying `Polymake.jl` object from a `Polyhedron`, respectively wrapping a `Polymalke.jl` object representing a polyhedron into a high-level `Polyhedron` object.
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
````diff
@@ -0,0 +1,55 @@
+pm_polytope(::Polyhedron)
+```
+
+The following shows all the data "known to" a `Polyhedron`.
````
I don't see how this sentence fits with the example coming next?

lkastner replied:
Please look at the output. https://oscar-system.github.io/Oscar.jl/previews/PR611/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake/ It displays all properties that polymake already computed of the cube. Probably the wording should be different.

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
```diff
@@ -0,0 +1,55 @@
+
+`polymake` allows for an interactive visualization of 3-dimensional polytopes in the browser: `Polymake.visual(C.pm_polytope)`.
+
+There are several design differences between `polymake` and `Oscar`.
```
... And... ? Are you going to tell us? Or is the "warning" coming next part of this message? If so, it should be made clear, e.g.:
```suggestion
There are several design differences between `polymake` and `Oscar`. The most important are the following:
```
But actually, it might be better to either ditch this sentence, or move it into that "warning"...

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
```diff
@@ -0,0 +1,55 @@
+	Indices in `polymake` are zero-based, whereas in `Oscar` they are one-based.
+
+The next example shows a purely combinatorial construction of a polytope (here: a square).
+Inspite of being given no coordinates, we can see that this is a simple polytope; i.e., each vertex is contained in dimension many facets.
```
```suggestion
In spite of being given no coordinates, `polymake` can check for us that this is a simple polytope; i.e., each vertex is contained in dimension many facets.
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_polymake.md:
````diff
@@ -0,0 +1,55 @@
+Q.SIMPLE
+```
+
+However, without coordinates, e.g., computing the volume cannot work:
````
```suggestion
However, without coordinates, some operations such as computing the volume cannot work:
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_serialization.md:
```diff
@@ -0,0 +1,27 @@
+
+# Saving and loading
+
+Objects of type Polyhedron can be saved to a file and loaded from a file in the
```
```suggestion
Objects of type `Polyhedron` can be saved to a file and loaded from a file in the
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_serialization.md:
````diff
@@ -0,0 +1,27 @@
+s = load_polyhedron("square.poly")
+s == square
+```
+The file is in json format and contains all the underlying polymake object. In
````
```suggestion
The file is in JSON format and contains all the underlying polymake object. In
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_cones.md:
```diff
@@ -12,6 +12,36 @@ Pages = ["pg_cones.md"]
+A set $C \subseteq \mathbb{F}^n$ is called a (polyhedral) cone if it can be written as the set of nonnegative linear combinations of finitely many vectors in $\mathbb{F}^n$.
+Equivalently, cones can be written as the intersection of finitely many homogeneous linear inequalities.
+
+Any cone is a special case of a polyhedron.
```
Not part of this review, but: If a cone is a polyhedron, then should perhaps `Cone` be a specialization of `Polyhedron`, or at least some kind of common parent type `AbstractPolyhedron`? (Another subtype then might be `CombinatorialPolyhedron`)

fingolfin commented on docs/src/PolyhedralGeometry/pg_cones.md:
```diff
@@ -12,6 +12,36 @@ Pages = ["pg_cones.md"]
+Any cone is a special case of a polyhedron.
+Conversely, intersecting a cone with a suitable affine hyperplane yields a polyhedron whose faces are in bijection with the faces of the cone.
+Going back and forth between polyhedra and their homogenizations, the cones, is a frequent operation.
+This is one reason for keeping cones as a distinct type.
```
That doesn't really make it clearer to me. I am not doubting there are good reasons, it's just this sentence doesn't make it clear to me; so I'd say either elaborate, or else ditch this?

fingolfin commented on docs/src/PolyhedralGeometry/pg_cones.md:
````diff
@@ -12,6 +12,36 @@ Pages = ["pg_cones.md"]
+One can construct a cone through an interior description, analogous  to the $V$-description of a polytope. By providing a collection of vectors, the command `positive_hull` will construct the cone generated by those vectors.
+
+```@docs
+positive_hull(::Union{MatElem, AbstractArray{T,2} where T})
````
We recently switched to using the type aliases `Vector` and `Matrix` whenever possible instead of `Array`, please adapt to that here, too.
```suggestion
positive_hull(::Union{MatElem, AbstractMatrix{T} where T})
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_cones.md:
````diff
@@ -12,6 +12,36 @@ Pages = ["pg_cones.md"]
+CC = load_cone("C.cone")
+collect(rays(CC))
+```
+The file is in json format and contains all the underlying polymake object. In
````
```suggestion
The file is in JSON format and contains all the underlying polymake object. In
```

fingolfin commented on src/Polytopes/Cone/properties.jl:
```diff
@@ -114,12 +102,9 @@ julia> ambient_dim(C)
 
-# Arguments
-- `C::Cone`: A cone.
+Returns the codimension of `C`.
```
```suggestion
Return the codimension of `C`.
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -53,9 +62,9 @@ end
+    objective_function(LP::LinearProgram; as = :pair)
 
-Returns the objective function x ↦ dot(c,x)+k of the linear program lp.
+Returns the objective function x ↦ dot(c,x)+k of the linear program LP.
```
```suggestion
Return the objective function x ↦ dot(c,x)+k of the linear program LP.
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -73,6 +82,12 @@ function objective_function(lp::LinearProgram; as::Symbol = :pair)
+    feasible_region(lp::LinearProgram)
+
+Returns the feasible region of the linear program lp, which is a Polyhedron.
+
```
```suggestion
Return the feasible region of the linear program lp, which is a `Polyhedron`.
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
````diff
@@ -82,7 +97,34 @@ feasible_region(lp::LinearProgram) = lp.feasible_region
+The following example constructs a linear program over the three dimensional cube.
+Although the linear program is given using the `:max` convention, one may still call
+`minimal_vertex`.
+```julia-repl
````
These also should all be turned into `jldoctest` if possible.

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -126,7 +246,7 @@ end
 ###############################################################################
 ###############################################################################
-
+#=
```
Why are you commenting out this code? If you don't need it anymore, just remove it?

lkastner committed:
Polytopes: Change show function back, but catch polytopes without embedding

lkastner committed:
Polytopes: julia-repl -> jldoctest

- Repair documentation examples to pass jldoctest
- Supress comparison of output for hilbert_basis, since it can come in
  various permutations from polymake

lkastner committed:
Polytope: Apply suggestions from code review

Those were obvious errors and typos

Co-authored-by: Max Horn <max@quendi.de>
Co-authored-by: Benjamin Lorenz <benlorenz@users.noreply.github.com>

lkastner committed:
It seems that these lines were removed by accident.

lkastner committed:
Polytopes: Apply suggestions from code review

Co-authored-by: Max Horn <max@quendi.de>

lkastner committed:
Polytopes: description -> representation in docs

lkastner committed:
Polytopes: Move sentence into warning as suggested

lkastner committed:
Merge branch 'docs/Polytopes' of github.com:oscar-system/Oscar.jl into docs/Polytopes

Commit:
Merge remote-tracking branch 'origin/master' into docs/Polytopes

Commit:
Changed the list layout of possible stings for Archimedean solids into Markdown

lkastner committed:
Polytopes: full dim -> full-dim

lkastner committed:
Polytopes: Returns -> Return

lkastner committed:
Polytopes: Work on Max comments from PR508

lkastner committed:
Polytopes: Wrapping lines in docs as requested

lkastner committed:
Polytopes: Rewording awkward sentence as suggested

lkastner committed:
Polytopes: Some more line wrapping in docs, some emph for definitions

lkastner committed:
Polytopes: Remove unused code in LinearProgram

lkastner committed:
Polytopes: Apply suggestions from code review

Co-authored-by: Max Horn <max@quendi.de>

lkastner committed:
Polytopes: Adhere to convention

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,88 @@
+
+## H and V representations
+
+### Intersecting halfspaces: H-representation
```
```suggestion
### Intersecting halfspaces: $H$-representation
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
````diff
@@ -0,0 +1,88 @@
+Polyhedron(A::Union{Oscar.MatElem,AbstractMatrix}, b)
+```
+
+### Computing convex hulls: V-representation
````
```suggestion
### Computing convex hulls: $V$-representation
```

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,88 @@
+constructed through other means: by name, via operations on other polyhedra, or
+from other objects in Oscar.
+
+## H and V representations
```
```suggestion
## $H$- and $V$-representations
```

fingolfin replied:
Or just drop this intermediate headline? It feels a bit redundant. But of course that's a matter of taste, so feel free to keep it in.

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md:
```diff
@@ -0,0 +1,88 @@
+## Polyhedra from other objects
+
+There is a growing list of functionality for constructing polyhedra coming from
+other mathematical objects. We list a handful below.
```
This would be fine for a presentation, a demo, a press release, but not a reference manual. Here, we talk about what is, not what might be (as "growing" does), and we strive to list all relevant functions we provide, not just "a handful".

So I'd remove this.

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_intro.md:
```diff
@@ -0,0 +1,36 @@
+
+# Introduction
+
+Let $\mathbb{F}$ be an ordered field; the most prominent case here is where
```
I still don't understand in what sense this is a "prominent" case? Do you mean "the most common" or "most frequently used", or "typically this is..."

fingolfin replied:
OK, something weird is going on here: I now see that you answered this question at https://github.com/oscar-system/Oscar.jl/pull/611#discussion_r686700104 but I cannot add to that conversation, nor did it show it to me in the code.. Huh.

Based on what @lkastner wrote there, I'd suggest to basically use his wording from his explanation (i.e., this is most frequently used, and for other fields not all features may be supported -- although of course ideally we'd be more specific as to what works and what not, but that can be in a future PR)

micjoswig replied:
While I am not picky about the wording, we need to compromise a bit here: Currently we only support rational polytopes etc in Oscar, but polymake can do much more.  The relevant ordered fields (e.g., rational functions) exist in Flint etc, but we need some glue code (which is in the making).  Therefore I'd prefer this slightly more general description to sketch our intentions.

micjoswig replied:
> I still don't understand in what sense this is a "prominent" case? Do you mean "the most common" or "most frequently used", or "typically this is..."

We could take "default".

fingolfin commented on src/Polytopes/Polyhedron/properties.jl:
```diff
@@ -89,21 +99,17 @@ function faces(as::Type{T}, P::Polyhedron, face_dim::Int) where {T}
         PolyhedronFacePolyhedronIterator(P,face_dim-size(lineality_space(P),1))
     else
-        throw(ArgumentError("Unsupported `as` argument :" * string(as)))
+        throw(ArgumentError("Unsupported `as` argument:" * string(as)))
```
In the future, I'd suggest to use string interpolation in this error messages (but this can be in another PR)
```suggestion
        throw(ArgumentError("Unsupported `as` argument: $(as)"))
```

micjoswig replied:
By the way, I have no idea how these threads of discussions are meant to converge.

fingolfin reviewed (approved):
Overall, this is much better now. There are still a few minor things I'd like to see adjusted, but e.g. the section about the underlying polymake/Polymake.jl stuff could of course be adjusted in a future PR (possibly taking my comments here into account)

micjoswig committed:
Update docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md

Co-authored-by: Max Horn <max@quendi.de>

micjoswig committed:
Update docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_constructions.md

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan commented on src/Polytopes/Polyhedron/constructors.jl:
```diff
@@ -60,43 +58,56 @@ pm_polytope(P::Polyhedron) = P.pm_polytope
-    convex_hull(V [, R [, L]])
+    convex_hull(V::Matrix [, R::Matrix [, L::Matrix]]; non_redundant::Bool = false)
+
+Compute the convex hull of the rows of the matrix `V`. Optionally one can give
```
Julia documenting conventions suggests to leave out redundancy where possible. This means not to state in the description that optional arguments are optional, if we can make this visible within the header in the docstring. Here, the square brackets do this job.

alexej-jordan committed:
applied suggestion from github comments

alexej-jordan reviewed (approved):
This looks really good for a presentable and usable version of the docs! The small issues that remain can be adjusted when this project proceeds.

lkastner committed:
Polytopes: Add references to polymake and Polymake.jl in docs

lkastner committed:
Polytopes: Deal with another batch of comments

lkastner committed:
Polytopes: Repair warning, use string interpolation as suggested

lkastner committed:
Polytopes: Incorporate more comments from PR508

lkastner committed:
Polytopes: Typo

lkastner committed:
Polytopes: Dummy types rewritten

lkastner committed:
Polytopes: Reword example to closer reflect output

lkastner committed:
Polytopes: generators -> R in positive_hull, elaborate more

lkastner committed:
Polytopes: Remnants of PR611 comments

fingolfin commented on docs/oscar_references.bib:
```diff
@@ -501,3 +501,34 @@ @book{Marcus
+   publisher="Springer International Publishing",
+   address="Cham",
+   pages="377--385",
+   abstract="We present the Julia interfaceto polymake, a software for research in polyhedral geometry. We describe the technical design and how the integration into Julia makes it possible to combine polymake with state-of-the-art numerical software.",
```
```suggestion
   abstract="We present the Julia interface to polymake, a software for research in polyhedral geometry. We describe the technical design and how the integration into Julia makes it possible to combine polymake with state-of-the-art numerical software.",
```

lkastner replied:
Thanks for catching this. This is actually the original bibtex from the Springer page, amazing.

lkastner committed:
Polytopes: Typo in reference to Polymake.jl

fingolfin commented on docs/src/PolyhedralGeometry/Polyhedra/pg_polyhedra_serialization.md:
````diff
@@ -0,0 +1,26 @@
+s = load_polyhedron("square.poly")
+s == square
+```
+The file is in JSON format and contains all the underlying polymake object. In
````
Something is off about the grammar here... perhaps like this? (Of course there are multiple places repeating the same phrase...)
```suggestion
The file is in JSON format and contains all the underlying polymake objects. In
```

alexej-jordan replied:
I think what the author meant was that it contains all (known/already computed) information of the polymake object so no re-computation is needed when loading the object for the price of bigger file size. I will re-phrase it so it is more clear.

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
```diff
@@ -10,11 +10,109 @@ using Oscar
 
 ## Introduction
 
+The task of a linear program is to optimize a linear function over a
```
```suggestion
The purpose of a linear program is to optimize a linear function over a
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
```diff
@@ -10,11 +10,109 @@ using Oscar
+
 ## Constructions
 
+Linear programs are constructed by passing a polyhedron, and a linear function
```
Perhaps introduce the term "objective function" here already, that might help connect the text with the example code for readers who are not experts on this kind of math?
```suggestion
Linear programs are constructed from a polyhedron and a linear *objective function*
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
```diff
@@ -10,11 +10,109 @@ using Oscar
 ## Constructions
 
+Linear programs are constructed by passing a polyhedron, and a linear function
+in terms of a vector and translation. One may specify whether the optimization
```
```suggestion
which is described by a vector and (optionally) a translation. One can select whether the optimization
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
```diff
@@ -10,11 +10,109 @@ using Oscar
 
+Linear programs are constructed by passing a polyhedron, and a linear function
+in terms of a vector and translation. One may specify whether the optimization
+problem is in terms of maximum or minimum.
```
```suggestion
problem is to maximize or to minimize the objective function.
```
Perhaps also state what the default is, minimize or maximize?

alexej-jordan replied:
For my feeling it is okay here to leave the default out because the docstring follows right after and should contain this information in its header.

I still agree with the improved phrasing.

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
````diff
@@ -10,11 +10,109 @@ using Oscar
+The information about the linear program `LP` can be easily extracted.
+
+```@repl oscar
+(c,k)=objective_function(LP)
````
```suggestion
c, k = objective_function(LP)
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
````diff
@@ -10,11 +10,109 @@ using Oscar
+`v`.
+
+```@repl oscar
+(m,v) = solve_lp(LP)
````
```suggestion
m, v = solve_lp(LP)
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
```diff
@@ -10,11 +10,109 @@ using Oscar
+
+Solving the optimization problem can be achieved using `solve_lp`: the output
+is a pair `(m,v)` where the optimal value is `m`, and that value is achieved at
+`v`.
```
```suggestion
To solve the optimization problem call `solve_lp`, which returns a pair `m, v`
where the optimal value is `m`, and that value is achieved at `v`.
```

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
````diff
@@ -10,11 +10,109 @@ using Oscar
+```
+
+Calling `solve_lp` on a linear program outputs a pair: the optimal value and
+the vertex at which the optimum is obtained
````
This paragraph again seems redundant: this info should be in the docstring for `solve_lp`; and perhaps also in a high-level description of the LP capabilities, but that was already given above...

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
```diff
@@ -10,11 +10,109 @@ using Oscar
+## Functions
+
+After constructing a linear program, it is straightforward to extract the
+feasible region and objective function
```
Redundant filler text: remove

fingolfin commented on docs/src/PolyhedralGeometry/pg_linear_programs.md:
````diff
@@ -10,11 +10,109 @@ using Oscar
+minimal_value
+```
+
+One can also obtain the maximal or minimal vertex at which the objective
````
First off, shouldn't it be "a maximal/minimal" vertex (it won't be unique in general)?

Secondly: I already said what I think about all this filler text, soooo I'll just try to not bring it up again, just imagine I've said it about all the other similar filler text here and elsewhere ;-).

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -73,6 +82,11 @@ function objective_function(lp::LinearProgram; as::Symbol = :pair)
+
+Return the feasible region of the linear program lp, which is a `Polyhedron`.
+"""
 feasible_region(lp::LinearProgram) = lp.feasible_region
```
Just wondering: you changed `lp` to `LP` in many places, but not all. Is that intentional? I mean, it is fine by me, just wanted to draw your attention to it. It's fine by me to leave it as-is (and/or address it in a future PR), just sayin'.

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -73,6 +82,11 @@ function objective_function(lp::LinearProgram; as::Symbol = :pair)
+"""
+    feasible_region(lp::LinearProgram)
+
+Return the feasible region of the linear program lp, which is a `Polyhedron`.
```
```suggestion
Return the feasible region of the linear program `lp`, which is a `Polyhedron`.
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -82,7 +96,34 @@ feasible_region(lp::LinearProgram) = lp.feasible_region
+Return, if it exists, a point of the feasible region of `LP` which minimizes the objective
+function of `LP`.
+
+Return `nothing` if the objective function does not obtain a minimal value over the feasible region.
```
```suggestion
Return either a point of the feasible region of `LP` which minimizes the objective
function of `LP`, or `nothing` if no such point exists.
```
If you prefer to keep your current text, please change "obtain" to "attain".

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -101,14 +169,65 @@ function maximal_vertex(lp::LinearProgram)
+    minimal_value(LP::LinearProgram)
+
+Return, if it exists, the minimal value of the objective function of LP over the feasible region
+   of LP. Otherwise, return `-inf`.
```
```suggestion
Return, if it exists, the minimal value of the objective function of `LP` over the feasible region
of `LP`. Otherwise, return `-inf`.
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -101,14 +169,65 @@ function maximal_vertex(lp::LinearProgram)
+    maximal_value(LP::LinearProgram)
+
+Return the maximal value of the objective function of LP over the feasible region
+   of LP. Otherwise, return `inf`
```
```suggestion
Return the maximal value of the objective function of `LP` over the feasible region
of `LP`. Otherwise, return `inf`
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -101,14 +169,65 @@ function maximal_vertex(lp::LinearProgram)
 
 Gives a pair `(m,v)` where the optimal value `m` of the objective
- function of lp is attained at `v` (if `m` exists). If the optimum
+ function of LP is attained at `v` (if `m` exists). If the optimum
```
```suggestion
 function of `LP` is attained at `v` (if `m` exists). If the optimum
```

fingolfin commented on src/Polytopes/LinearProgram.jl:
```diff
@@ -101,14 +169,65 @@ function maximal_vertex(lp::LinearProgram)
-    solve_lp(lp::LinearProgram)
+    solve_lp(LP::LinearProgram)
 
 Gives a pair `(m,v)` where the optimal value `m` of the objective
```
```suggestion
Return a pair `(m,v)` where the optimal value `m` of the objective
```

fingolfin commented on src/Polytopes/PolyhedralFan/constructors.jl:
```diff
@@ -5,19 +5,63 @@
+@doc Markdown.doc"""
+    PolyhedralFan(Rays, Cones)
+
+# Arguments
```
This should go (it's not "julia-ish"). But can be in another PR

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
````diff
@@ -22,23 +22,64 @@ Base.length(iter::PolyhedralFanRayIterator) = nrays(iter.fan)
+```jldoctest
+julia> PF = face_fan(cube(3));
+
+julia> for c in maximal_cones(PF)
+       println(nrays(c))
+       end
+4
+4
+4
+4
+4
+4
````
Maybe something likethis would be more Julia-ish?
```suggestion
julia> map(nrays, maximal_cones(PF))
6-element Vector{Int64}:
 4
 4
 4
 4
 4
 4
```

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -71,30 +112,65 @@ Base.length(iter::MaximalConeIterator) = nmaximal_cones(iter.PF)
 
-Returns the number of maximal cones in a polyhedral fan `PF`.
+Return the number of maximal cones of `PF`.
+# Examples
```
```suggestion

# Examples
```

fingolfin commented on src/Polytopes/PolyhedralFan/properties.jl:
```diff
@@ -103,27 +179,95 @@ nrays(PF::PolyhedralFan) = pm_fan(PF).N_RAYS
+
+# Examples
+The rays of a face fan (centered at the origin) point into the direction of the
+polytope's vertices; here we see this for the `3`-cube.
```
```suggestion
polytope's vertices; here we see this for the 3-cube.
```
At least that would match how it is written elsewhere. You can of course also change it everywhere else to use `3`-cube, but it seems a bit odd (I could understand `$3$-cube`, though, but I don't think there'll be a visual difference in the resulting HTML?)

fingolfin commented on src/Polytopes/Polyhedron/constructors.jl:
```diff
@@ -158,7 +172,12 @@ end
 function Base.show(io::IO, P::Polyhedron)
-    print(io, "A polyhedron")
+    ad = ambient_dim(P)
+    if ad == -1.0
```
Why is the ambient dimension compared to a float, is there some subtlety I am missing here?
```suggestion
    if ad == -1
```

fingolfin commented on src/Polytopes/Polyhedron/properties.jl:
```diff
@@ -639,20 +603,16 @@ end
 
-# Arguments
-- `P::Polyhedron`: A polyhedron.
+Return the ambient dimension of `P`.
```
In an example elsewhere, I learned that apparently this (or perhaps some other `ambient_dim` method??? not sure) can return -1 (or perhaps -1.0?) to indicate "there is no ambient dimension" (whatever that means). This ought to be mentioned here?

fingolfin commented on src/Polytopes/Polyhedron/properties.jl:
```diff
@@ -140,88 +144,82 @@ end
         VertexPointIterator(P)
     else
-        throw(ArgumentError("Unsupported `as` argument :" * string(as)))
+        throw(ArgumentError("Unsupported `as` argument:" * string(as)))
```
If you change one of them, perhaps change all? ;-).
```suggestion
        throw(ArgumentError("Unsupported `as` argument: $(as)"))
```

fingolfin commented on src/Polytopes/Polyhedron/properties.jl:
```diff
@@ -3,17 +3,28 @@
+"""
+    Points
+
+Empty type for selecting `Vector` as return type for points.
```
This is of course off-topic for this PR, but: I don't get why you need these two empty types. Why don't you just accept `Polyhedron` resp. `Vector` instead?

fingolfin reviewed (commented):
Here's a bunch of more remarks. Many (most?) are cosmetic, but there are a few you really may want to address. But I'll leave it up to you whether to do it now or in the future *shrug*.

fingolfin commented:
Also: good work, this is a major improvement on the quality of the docstrings and the reference manual, thanks!

alexej-jordan committed:
Update docs/src/PolyhedralGeometry/pg_linear_programs.md

phrasing

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update docs/src/PolyhedralGeometry/pg_linear_programs.md

phrasing (linear program)

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update docs/src/PolyhedralGeometry/pg_linear_programs.md

phrasing (linear program)

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update docs/src/PolyhedralGeometry/pg_linear_programs.md

phrasing (linear program)

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update docs/src/PolyhedralGeometry/pg_linear_programs.md

adhere to convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update docs/src/PolyhedralGeometry/pg_linear_programs.md

adhere convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/LinearProgram.jl

adhere to convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/LinearProgram.jl

phrasing (minimal point)

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/LinearProgram.jl

adhere to convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/LinearProgram.jl

adhere to convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/LinearProgram.jl

adhere to convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/LinearProgram.jl

adhere to convention

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
Update src/Polytopes/PolyhedralFan/properties.jl

adding vertical distance

Co-authored-by: Max Horn <max@quendi.de>

alexej-jordan committed:
phrasing and interpunction in linear program; phrasing for file saving

alexej-jordan committed:
phraasing (linear program)

alexej-jordan committed:
Update src/Polytopes/PolyhedralFan/properties.jl

consistency

Co-authored-by: Max Horn <max@quendi.de>
